### ISSUE TITLE
[SUGGESTION] Namespace definition to use `+=` instead of `=`

### ISSUE THREAD

**Issue by JohelEGP** [hsutter/cppfront#438]
> So after some thought of brainstorming options, include ones that include `is`, the one I like the best is to use exactly the declaration syntax, but with `==` to connote equality with an existing entity, instead of `=` which connotes setting the new entity's value.

A wild thought.

Maybe a namespace declaration should use `+=` instead of `=`.
So `ns: namespace = {` would become `ns: namespace += {`.

`+=` denotes in-place addition, not definition.
But notice how `:` remains despite `+=`.
So it both (re)defines the namespace's identifier, which is valid, and adds to it.
Then all remaining uses of `: /*signature*/ =` would denote definition of the non re-definable kind, which is a bigger plus.

How does that fit with the current Cpp2?
What about with your idea to replace `namespace` with a metafunction, @hsutter?

_Originally posted by @JohelEGP in https://github.com/hsutter/cppfront/issues/273#issuecomment-1537261489_

**Comment by JohelEGP** [comment]
> By supporting nested namespace definition, the distinction from other declarations raises.
> ```Cpp2
> x::y: type = { } // Error!
> a::b: namespace += { } // Perfectly OK.
> ```
> 
> _Originally posted by @JohelEGP in https://github.com/hsutter/cppfront/issues/273#issuecomment-1537694999_

> > > Then all remaining uses of : /_signature_/ = would denote definition of the non re-definable kind, which is a bigger plus.
> > 
> > Varibles can be re-assigned so I don't think `: /*signature*/ =` denotes something of non-redefinable kind and you can also add functions to classes.
> > _Originally posted by @AbhinavK00 in https://github.com/hsutter/cppfront/issues/273#issuecomment-1537860248_
>
> Variables can be assigned to, yes.
> But only once, the first time, may it have `:` indicating definition.
> Other assignments are not redefinitions.
> 
> Shadowing does redefine what the identifier means for the current scope.
> And is distinguished from `: namespace +=`, which really does append, and doesn't shadow.
> 
> > ```c++
> > // Maybe they should be defined at first with `=`
> > x: namespace = {}
> > ```
> 
> I disagree.
> Maybe that should be the case for the extension methods you describe.
> But `:` is definition, and namespaces always append, so requiring `=` first is just needless friction.
> _Originally posted by @JohelEGP in https://github.com/hsutter/cppfront/issues/273#issuecomment-1538081839_

**Comment by JohelEGP** [comment]
> Shadowing does redefine what the identifier means for the current scope.
> And is distinguished from `: namespace +=`, which really does append, and doesn't shadow.

Maybe the last "and doesn't shadow" is wrong.
It's still a definition, so it does introduce a name.
```Cpp2
x: int = 0;
f: () = {
  x := 0; // Unqualified `x` no longer means `::x`.
}
a: namespace += { }
b: namespace += {
  a: namespace += { } // Unqualified `a` now means `b::a`.
}
```

**Comment by msadeqhe** [comment]
I like your idea, `+=` is more expressive than `=` about what happens when we declare namespaces.

> Maybe the last "and doesn't shadow" is wrong.
> It's still a definition, so it does introduce a name.

Because shadowing happen on declarations, so `=` or `+=` don't affect shadowing but `:` does.

**Comment by gregmarr** [comment]
@AbhinavK00 Yes, that's true, and Johel is suggesting that this "same syntax" for namespaces be `+=` instead of `=`, just like the syntax for aliases is `==` instead of `=`, to point out that it's not exactly the same as a normal one-shot definition.  (I'm not saying whether I agree with it or not, just clarifying the proposal.)

**Comment by AbhinavK00** [comment]
Ah ok, so here's what I think.
`+=` has no advantage over `=`. `=` is more consistent because it is also used with functions, variables and classes, moreover none of these has a valid `: /*identifier*/ +=` syntax. Using `+=` loses that consistency. I know it is inspired from `==` for aliases but if we went this path, then the syntax would just diverge into`name : identifier "something out of =, == or += depending on identifier` instead of same `:=` for all the things. Yes, aliases are an exception but maybe they would be fixed in future if we come up with a better syntax that is context free

**Comment by JohelEGP** [comment]
It's already context free.

**Comment by realgdman** [comment]
https://github.com/hsutter/cppfront/wiki/Design-note:-Namespaces says namespaces could become partial classes.

**Comment by JohelEGP** [comment]
Yes. That's what this from the OP was alluding to:

> What about with your idea to replace `namespace` with a metafunction, @hsutter?

**Comment by msadeqhe** [comment]
@AbhinavK00, Good point about consistency :+1:. So for this reason, still `: =` seems natural for namespace declarations.

> ...
> I'm not talking about cpp2 has a whole. Someone on reddit recommended the following syntax for aliases
> 
> ```
> x : alias = something;
> ```
> 
> and Herb pointed that this won't be context free as the compiler would have to look up whether "something" is a class, namespace or else. The current syntax is context free as you specify upfront what you're aliasing to.
> ...

IMO aliases don't define/implement/initialize/... anything, they are just alternative names. So they don't need to be the same as other declarations which define/implement/initialize/... new types/functions/variables/... .
